corehq/form_processor/backends/couch/processor.py

```diff
@@ -81,6 +84,47 @@ def hard_delete_case_and_forms(cls, domain, case, xforms):
         docs = [case._doc] + [f._doc for f in xforms]
         case.get_db().bulk_delete(docs)
 
+    # TODO: add tests
+    @classmethod
+    def update_responses(cls, xform, value_responses_map, user_id):
```

Comment:
There's quite a lot of overlap between the couch and sql versions of this method. I think a lot could be moved into the `FormProcessorInterface` class. Something like this:

```python
# in FormProcessorInterface
def update_responses(xform, value_responses_map, user_id)
    existing_form = FormAccessors(self.domain).get_with_attachments(xform.form_id)
    xml = existing_form.get_xml_element()
    dirty = False
    for question, response in value_responses_map.iteritems():
        if update_response(xml, question, response, xmlns=xform.xmlns):
            dirty = True

    if dirty:
        # slightly different for couch and sql hence delegated to the specifc backend processor
        new_form = self.processor.new_form_from_old(existing_form)

        existing_form, new_form = apply_deprecation(existing_form, new_form)
        new_xml = etree.tostring(xml)
        self.store_attachments(new_form, [
            Attachment(name=ATTACHMENT_NAME, raw_content=new_xml, content_type='text/xml')
        ])
        self.save_processed_models([new_form, existing_form])
```

(sorry got a bit carried away there)

Reply:
This is great, thank you! Not having worked with this code much, I appreciate all the details.